File: backend/nodejs-typescript/src/modules/auth/auth.service.ts
import bcrypt from "bcrypt";
import { LoginDto, RegisterDto } from "./auth.schema";
import User from "@/models/user";
import Session from "@/models/session";
import moment from "@/libs/moment";
import {
  ValidationError,
  AuthorizationError,
} from "@/utils/response-error.util";

export default class AuthService {
  static async register(registerDto: RegisterDto) {
    await User.transaction(async (trx) => {
      const existingUser = await User.query(trx)
        .where("email", registerDto.email)
        .orWhere("phone_number", registerDto.phoneNumber)
        .first()
        .forUpdate();

      if (existingUser) {
        throw new ValidationError("Email or phone number already used");
      }

      const hashedPassword = await bcrypt.hash(registerDto.password, 10);

      await User.query(trx).insert({
        name: registerDto.name,
        email: registerDto.email,
        phone_number: registerDto.phoneNumber,
        password: hashedPassword,
        bio: registerDto.bio,
      });
    });

    return {
      message: "User registered successfully",
    };
  }

  static async login(loginDto: LoginDto) {
    const user = await User.query()
      .where("email", loginDto.emailOrPhoneNumber)
      .orWhere("phone_number", loginDto.emailOrPhoneNumber)
      .first();

    if (!user) {
      throw new AuthorizationError(
        "Email, phone number, or password is incorrect",
      );
    }

    const isPasswordValid = await bcrypt.compare(
      loginDto.password,
      user.password,
    );

    if (!isPasswordValid) {
      throw new AuthorizationError(
        "Email, phone number, or password is incorrect",
      );
    }

    const session = await Session.query().insert({
      user_id: user.id,
      expires_at: moment().add(1, "day").unix(),
    });

    return {
      message: "User logged in successfully",
      data: {
        sessionId: session.id,
        expiresAt: session.expires_at,
      },
    };
  }
}

File: backend/go/internal/services/auth_service.go
package services

import (
	"time"

	"github.com/fathurmdr/backend/go/internal/db"
	"github.com/fathurmdr/backend/go/internal/dto"
	"github.com/fathurmdr/backend/go/internal/models"
	"github.com/fathurmdr/backend/go/internal/utils"
	"golang.org/x/crypto/bcrypt"
)


type AuthService struct{}

func (authService *AuthService) Register(registerRequest *dto.RegisterRequest) (*models.User, error) {
	var existingUser models.User
	err := db.DB.Where("email = ? OR phone_number = ?", registerRequest.Email, registerRequest.PhoneNumber).First(&existingUser).Error
	if err == nil {
		return nil, utils.NewValidationError("Email or phone number already used", nil)
	}

	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(registerRequest.Password), bcrypt.DefaultCost)
	if err != nil {
		return nil, utils.NewApplicationError("", nil)
	}

	user := &models.User{
		Name:        registerRequest.Name,
		Email:       registerRequest.Email,
		PhoneNumber: registerRequest.PhoneNumber,
		Password:    string(hashedPassword),
		Bio:         registerRequest.Bio,
	}
	err = db.DB.Create(&user).Error
	if err != nil {
		return nil,  utils.NewApplicationError("", nil)
	}

	return user, nil
}

func (authService *AuthService) Login(loginRequest *dto.LoginRequest) (map[string]interface{},  error) {
	var user models.User
	err := db.DB.Where("email = ? OR phone_number = ?", loginRequest.EmailOrPhoneNumber, loginRequest.EmailOrPhoneNumber).First(&user).Error
	if err != nil {
		return nil, utils.NewAuthorizationError("Email, phone number, or password is incorrect", nil)
	}

	err = bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(loginRequest.Password))
	if err != nil {
		return nil, utils.NewAuthorizationError("Email, phone number, or password is incorrect", nil)
	}

	session := models.Session{
		UserID:    user.ID,
		ExpiresAt: time.Now().Add(24 * time.Hour).Unix(),
	}
	err = db.DB.Create(&session).Error
	if err != nil {
		return nil, utils.NewApplicationError("", nil)
	}

	return map[string]interface{}{
		"sessionId": session.ID,
		"expiresAt": session.ExpiresAt,
	}, nil
}
